### PR TITLE
Interface

### PR DESCRIPTION
**Pull request by nicholasunderwood**
Unit and math libs

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -13,8 +13,8 @@ repos:
     -   id: pretty-format-json
         args: [--autofix]
     - id: trailing-whitespace
--   repo: https://github.com/team5419/pre-commit-hooks
-    rev: v1.0.0
+-   repo: https://github.com/team5499/pre-commit-hooks
```
Comment (FelixMo42) [inline .pre-commit-config.yaml]: Shouldn't this be team 5419?

File: build.gradle [hunk 2]
```diff
@@ -2,12 +2,13 @@ import org.gradle.api.tasks.options.Option
 
 plugins {
     id 'org.jetbrains.kotlin.jvm' version '1.3.20'
-    id "edu.wpi.first.GradleRIO" version "2019.4.1"
+    id "edu.wpi.first.GradleRIO" version "2019.3.2"
```
Comment (FelixMo42) [inline build.gradle:5]: Why are we reverting to an older version?

File: src/main/kotlin/org/team5419/fault/auto/ActionGroup.kt [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    override fun finish() {
+        actionMap.forEach {
+            if (!it.value) it.key.finish()
+        }
+    }
+}
+
+open class SerialAction(actions: MutableList<Action>) : ActionGroup() {
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/auto/ActionGroup.kt:55]: Why is this a mutable list, that seems like a recipe for bugs?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (FelixMo42) [inline .pre-commit-config.yaml:16]: Shouldn't this be team 5419?

File: src/main/kotlin/org/team5419/fault/auto/ActionUtil.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+enum class GroupType { Parallel, Serial }
+
+interface IActionGroupBuilder {
+
+    fun build(): ActionGroup
+}
+
+class ActionGroupBuilder(private val type: GroupType) : IActionGroupBuilder {
+    private val actions = mutableListOf<Action>()
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/auto/ActionUtil.kt:11]: Again, don't use mutable list if you don't have to.

File: src/main/kotlin/org/team5419/fault/hardware/AbstractBerkeliumEncoder.kt [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+import org.team5419.fault.math.units.SIKey
+import org.team5419.fault.math.units.SIUnit
+import org.team5419.fault.math.units.derived.Velocity
+import org.team5419.fault.math.units.native.NativeUnitModel
+
+abstract class AbstractBerkeliumEncoder<T : SIKey>(
+    val model: NativeUnitModel<T>
+) : BerkeliumEncoder<T> {
+    override val position: SIUnit<T> get() = model.fromNativeUnitPosition(rawPosition)
+    override val velocity: SIUnit<Velocity<T>> get() = model.fromNativeUnitVelocity(rawVelocity)
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/hardware/AbstractBerkeliumEncoder.kt:12]: Shouldn't the interface include a way to zero the encoder?

File: src/main/kotlin/org/team5419/fault/hardware/AbstractBerkeliumMotor.kt [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+abstract class AbstractBerkeliumMotor<T : SIKey> : BerkeliumMotor<T> {
+
+    override var useMotionProfileForPosition = false
+
+    override fun follow(motor: BerkeliumMotor<*>): Boolean {
+        println("Cross brand following not implemented yet!")
+        return false
+    }
+}
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/hardware/AbstractBerkeliumMotor.kt:13]: We use a lot more functionality in the BerkeliumMotor class than just the follow function. It should be here

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -13,8 +13,8 @@ repos:
...
+    rev: v0.3.2
```
Comment (andycate) [inline .pre-commit-config.yaml:17]: ????

File: .travis.yml [hunk 8]
```diff
@@ -2,19 +2,14 @@ sudo: required
...
+  - secure: H4xBumrS4Z8OXqW+evg/iczS0kRaK7g/HAn7g+a4bwxfHYME6CN7hvwy9MY+McidDcXz/BD88eJnZ4/SyKUdlFFd5PqN5tiUGkHsY9QcQjRTw5alNgONk8S5vHDXqlRizF5EKUm7yBC4DRgFlmhF4TJybxwjI3dSFrKYakvEEtlCC6O3JuXq/PYdoz95AbmoxVrqkGwAojDZdhZ6UggeN+IJYi1h1FZGA0mD5PMKUKAHKxl5WVy9cQYV/+cc4Qy4c/0mrPBZ4KRbHA+UWxEnA/H9DARWLSQsDMZy7QCcQ0OXuM21kVVPt2qNE8wZQ074oX+MKcJbnvGT+AJpDPJwL+iRjch8SuDVpLOWE/LDcNE+r3PjUuL6HdA7/NBN0jhOsKTB3N/D3ehRGpkWlwXnK3zw9/P7HJ6I9MXPf4hq89B+Rg9vRdLAhVxFIuTOLMONVJyALrzX4sm7TaJAdfNC0NIBZpwWKmHzN3eVQovapUfEPo+no2cMXOxwyV/tKXeg6Hf+u4PvHNcmmyCnEDLNd4syayYHWExh4nHkGQPAa6fpQ7FNUdo0AkvwK6lU5ylpHsbsjcWOgWQ7vjbuu08mJ3JsNgb3Lo3kNSm9PXECC9M4rIjyUzTGrFZsS2GbtaS9M0txO1B8Ysv59Y3wruMZ89vpeOoZI6gJoHkLvU7FvYo=
```
Comment (andycate) [inline .travis.yml:15]: We need to fix this

File: README.md [hunk 9]
```diff
@@ -1,6 +1,4 @@
-# fault
-
-[![Build Status](https://travis-ci.org/team5419/fault.svg?branch=master)](https://travis-ci.org/team5419/fault)
-[![Code Coverage](https://codecov.io/gh/team5419/fault/branch/master/graph/badge.svg)](https://codecov.io/gh/team5419/fault)
-
-fault is team 5419's base kotlin code used in all of our robots. Mostly contains math and utilities
+# Fault
+
+My own branch of Fault
+
```
Comment (andycate) [inline README.md:4]: BRuH

File: build.gradle (hunk 2 above)
Comment (andycate) [inline build.gradle:5]: Noyan, use the right version

File: src/main/kotlin/org/team5419/fault/math/geometry/Rotation2d.kt [hunk 10]
```diff
@@ -1,120 +1,58 @@
 package org.team5419.fault.math.geometry
 
-import org.team5419.fault.math.Epsilon
+import org.team5419.fault.math.epsilonEquals
+import org.team5419.fault.math.kEpsilon
+import org.team5419.fault.math.units.SIUnit
+import org.team5419.fault.math.units.derived.*
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/math/geometry/Rotation2d.kt:6]: Isn't doing mass imports considered bad practice? I know are best friend Tox doesn't like it.

File: build.gradle [hunk 11]
```diff
@@ -32,10 +33,12 @@ dependencies {
     testCompile 'org.powermock:powermock-module-junit4:1.6.6'
     testCompile 'org.powermock:powermock-api-mockito2:1.6.6'
 
+    testCompile group: 'org.knowm.xchart', name: 'xchart', version: '3.5.4'
+
 }
 
 group = 'org.team5419'
-version = '2.11.1'
+version = '3.0.0-SNAPSHOT'
```
Comment (andycate) [inline build.gradle:41]: make sure to change this before merging

File: build.gradle [hunk 12]
```diff
@@ -111,7 +114,7 @@ publishing {
             artifact sourcesJar
             artifact javadocJar
             pom {
-                name = 'fault'
+                name = 'Fault'
```
Comment (andycate) [inline build.gradle:117]: change back

File: src/main/kotlin/org/team5419/fault/math/geometry/Vector2.kt [hunk 13]
```diff
@@ -1,68 +1,75 @@
+
 package org.team5419.fault.math.geometry
 
-@Suppress("TooManyFunctions")
-class Vector2(val x: Double, val y: Double) : Geometric<Vector2> {
+import org.team5419.fault.math.units.*
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/math/geometry/Vector2.kt:4]: Again, do we want to do mass imports?

File: src/main/kotlin/org/team5419/fault/subsystems/Subsystem.kt [hunk 14]
```diff
@@ -0,0 +1,45 @@
+package org.team5419.fault.subsystems
+
+import org.team5419.fault.util.loops.ILooper
+import java.util.concurrent.atomic.AtomicLong
+
+object SubsystemManager {
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/subsystems/Subsystem.kt:6]: Do we need a way to get subsystems? Are we just going to pass all of them around manually? How is this going to work?

**Review by andycate — CHANGES_REQUESTED** [review]
Fix some stuff

File: src/main/kotlin/org/team5419/fault/util/Source.kt [hunk 15]
```diff
@@ -0,0 +1,86 @@
+@file:Suppress("TooManyFunctions")
+package org.team5419.fault.util
+
+import kotlin.math.absoluteValue
+import kotlin.math.sign
+
+typealias Source<T> = () -> T
```
Comment (FelixMo42) [inline src/main/kotlin/org/team5419/fault/util/Source.kt:7]: What is this, the name doesn't say anything. This file need a bare minimum of docs.

**Review by FelixMo42 — COMMENTED** [review]
Have we done any benchmarking on if the SI unit system slows things down? It seems like that could be a real drag if it has to be constantly creating and destroying objects, but it might have some sort of internal optimization I don't know about.

File: src/main/kotlin/org/team5419/fault/auto/ActionGroup.kt (hunk 3 above)
Comment (nicholasunderwood) [reply src/main/kotlin/org/team5419/fault/auto/ActionGroup.kt:55]: I would have thought that would only cause problems with multiple threads

Comment (FelixMo42) [reply src/main/kotlin/org/team5419/fault/auto/ActionGroup.kt:55]: Actions are run sorta sudo multithreaded. We do one tick per loop, but that means that somewhere else in the update loop someone could still change the list.

Also, we are multithreading some stuff.

**Pull request closed without merging**